File: apps/zoe_app/lib/domain/entities/store.dart
/// Entidade Store do domínio.
///
/// Referência: ARCHITECTURE.md §ERD: STORES
class Store {
  final String id;
  final String name;
  final String slug;
  final String? logoUrl;
  final String? bannerUrl;
  final String? description;
  final double latitude;
  final double longitude;
  final double? distanceKm;
  final double rating;
  final bool isActive;
  final String? estimatedDeliveryTime;

  const Store({
    required this.id,
    required this.name,
    required this.slug,
    this.logoUrl,
    this.bannerUrl,
    this.description,
    required this.latitude,
    required this.longitude,
    this.distanceKm,
    this.rating = 0.0,
    required this.isActive,
    this.estimatedDeliveryTime,
  });

  factory Store.fromJson(Map<String, dynamic> json) {
    return Store(
      id: json['id'] as String,
      name: json['name'] as String,
      slug: json['slug'] as String? ?? '',
      logoUrl: json['logo_url'] as String?,
      bannerUrl: json['banner_url'] as String?,
      description: json['description'] as String?,
      latitude: (json['lat'] as num?)?.toDouble() ?? (json['latitude'] as num).toDouble(),
      longitude: (json['lng'] as num?)?.toDouble() ?? (json['longitude'] as num).toDouble(),
      distanceKm: (json['distance_km'] as num?)?.toDouble(),
      rating: (json['rating'] as num?)?.toDouble() ?? 0.0,
      isActive: json['is_active'] as bool? ?? true,
      estimatedDeliveryTime: json['estimated_delivery_time'] as String?,
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'name': name,
        'slug': slug,
        'logo_url': logoUrl,
        'banner_url': bannerUrl,
        'description': description,
        'latitude': latitude,
        'longitude': longitude,
        'distance_km': distanceKm,
        'rating': rating,
        'is_active': isActive,
        'estimated_delivery_time': estimatedDeliveryTime,
      };
}

File: backend/app/models/store.py
"""
Model: STORES

Referências:
- ARCHITECTURE.md §ERD: STORES { uuid id PK, text name, text cnpj, text status, timestamptz created_at }
- ARCHITECTURE.md §Store Entity: name, cnpj, description, logoUrl, bannerUrl, address, coordinates, rating
- TECHNICAL_AUDIT.md §1.3: coluna geohash para cache geoespacial otimizado
"""

import uuid
from datetime import datetime

from sqlalchemy import String, DateTime, Float, Integer, Boolean, Text, Enum as SAEnum
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database import Base

import enum


class StoreStatus(str, enum.Enum):
    ACTIVE = "active"
    INACTIVE = "inactive"
    SUSPENDED = "suspended"


class Store(Base):
    __tablename__ = "stores"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    cnpj: Mapped[str] = mapped_column(String(18), unique=True, nullable=False)
    description: Mapped[str] = mapped_column(Text, nullable=True)
    logo_url: Mapped[str | None] = mapped_column(String(512), nullable=True)
    banner_url: Mapped[str | None] = mapped_column(String(512), nullable=True)

    # Endereço embarcado
    address_street: Mapped[str] = mapped_column(String(255), nullable=False)
    address_city: Mapped[str] = mapped_column(String(100), nullable=False)
    address_state: Mapped[str] = mapped_column(String(2), nullable=False)
    address_zip: Mapped[str] = mapped_column(String(10), nullable=False)

    # Geolocalização (TECHNICAL_AUDIT §1.3 – PostGIS + geohash)
    lat: Mapped[float] = mapped_column(Float, nullable=False)
    lng: Mapped[float] = mapped_column(Float, nullable=False)
    geohash: Mapped[str | None] = mapped_column(String(12), index=True, nullable=True)

    rating: Mapped[float] = mapped_column(Float, default=0.0)
    review_count: Mapped[int] = mapped_column(Integer, default=0)
    status: Mapped[StoreStatus] = mapped_column(SAEnum(StoreStatus), default=StoreStatus.ACTIVE)
    is_open: Mapped[bool] = mapped_column(Boolean, default=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=datetime.utcnow)

    # Relationships
    products = relationship("Product", back_populates="store", lazy="selectin")
    shipments = relationship("OrderShipment", back_populates="store", lazy="selectin")
